instant-xml/src/lib.rs

```diff
@@ -10,15 +10,165 @@ pub use macros::{FromXml, ToXml};
 pub mod parse;
 
 pub trait ToXml {
-    fn write_xml<W: fmt::Write>(&self, write: &mut W) -> Result<(), Error>;
-
     fn to_xml(&self) -> Result<String, Error> {
-        let mut out = String::new();
-        self.write_xml(&mut out)?;
-        Ok(out)
+        let mut parent_prefixes = BTreeSet::new();
```

Comment:
I think we should keep the `W: fmt::Write` argument and put it in the `Serializer`. How painful is it if the `Serializer` has a generic type argument `W` for this? My experience from Askama is that virtual calls at this level can lead to pretty painful performance.

Reply:
I changed it, but I do not get the performance implications here, could you elaborate more about it? We can make a short call as well.

Reply:
If we have a generic `W: fmt::Write`, callers will monomorphize the implementations with a particular `fmt::Write` implementations. If we instead hold a `&mut dyn fmt::Write` in the `Serializer`, each call to write some data into the `fmt::Write` will have to go through the method table used for the `&dyn` fat pointer. Given that we're doing a lot of calls into that, the performance implications for that start to add up.